---
title: "Quality control and PCA for FLC1 RNA-seq"
author: "Edward Wallace"
date: "2024-12-04"
output: 
  html_document:
    toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(cowplot)
library(GGally)
library(here)
library(DESeq2)
library(biobroom)

theme_set(theme_cowplot(font_size = 12) +
    theme(# panel.grid.major = element_line(colour = "grey80", size = 0.5),
          strip.background = element_blank())
    )

# Count header columns from featurecounts
count_header_columns <- c("Geneid", "Chr", "Start",
                          "End", "Strand", "Length")

# unwanted string in filename; may need to be edited
unwanted_filename_string1 <- "11308_"
unwanted_filename_string2 <- "_S{[0-9]}+_R2_001_aln.bam"

```

# Summary

Quality control of RNA-seq data after read processing.

This analysis does:

- Scatter plot of counts showing log-scale correlations
- PCA plot showing how samples are similar to one another

It concludes that:

- Replicates group together in PCA space, so the experiment is reproducible
- PC1 contrasts WT with FLC1delta, capturing about 65% of variance
- PC2 contrasts temperature (for WT) and EGTA treatment (more for FLC1delta)
- There is a lot more to be understood here.

# Load and organise raw count data

## Load the raw counts

```{r load_counts}
counts_raw <- 
  here::here("results", "counts", "counts.txt") %>%
  readr::read_tsv(comment = "#")
counts_raw
```

## Load the sample sheet

```{r load_samplesheet}
sample_sheet <- 
  here::here("input", "experiment",
             "CryptoRNAseqFLC1SampleKey.csv") %>%
  readr::read_csv(comment = "#", col_types = "ccccccc") %>%
  dplyr::mutate(Condition = stringr::str_sub(SampleName,start = 1L, end = -3L))

sample_sheet 
```

## Organise the counts using the sample sheet

```{r organise_counts}
counts_all <- 
  counts_raw %>%
  dplyr::rename_with(stringr::word,
                     start = 2L, sep = fixed("_"), .cols = ends_with(".bam")) %>%
  dplyr::select(sample_sheet$SampleID)  %>%
  magrittr::set_colnames(sample_sheet$SampleName)  %>%
  magrittr::set_rownames(counts_raw$Geneid)

counts_all
```

## Write out counts with sample codes as column names

```{r write_counts_samplenames}
bind_cols(
  counts_raw %>% 
    dplyr::select(Gene = Geneid, Length = Length),
  counts_all 
) %>%
  readr::write_tsv(
    file = here::here("results", "summaries", "counts_bysamplecode.txt"))
```


# Check reproducibility by paired scatter plot in log10-plus-1-space

_Note: This is set to `eval=FALSE` as it can run very slowly for large datasets._ 

```{r paired_scatter, fig.height=20, fig.width=20, eval=FALSE}
counts_all %>%
  magrittr::add(1) %>%
  log10() %>%
  GGally::ggpairs()

here::here("results", "figures","paired_scatter_log10countsplusone_all_samples.png") %>%
ggsave(width = 32, height = 32)
```

# Normalize counts by regularized log (rlog) with DESeq2

The [rlog function from DESeq2](https://rdrr.io/bioc/DESeq2/man/rlog.html) "transforms the count data to the log2 scale in a way which minimizes differences between samples for rows with small counts, and which normalizes with respect to library size."

## Construct DESeq dataset object

EDIT: the `design` formula may need to be edited.

```{r dds_counts_all}
dds_all <- DESeqDataSetFromMatrix(countData = counts_all,
                                  colData = sample_sheet,
                                  design = ~ SampleName)
```

# Do the rlog transformation


```{r rlog_dds_all}
rlog_all <- rlog(dds_all)
head(assay(rlog_all))
```

# Check reproducibility by principal component analysis (PCA) of rlog

## Calculate PCs from the rlog output

Principal component analysis focused on comparing samples to other samples, hence take the transpose.

```{r calculate_pca_rlog}
# calculate principal components of rlog, after extracting from the dataset
pca_rlog <- rlog_all %>%
  assay() %>%
  t() %>%
  prcomp()

# convert principal components to data frame
pcdf_rlog <- bind_cols(
  as_tibble(colData(rlog_all)),
  as_tibble(pca_rlog$x)
)

pcdf_rlog
```

## Calculate proportion of variance explained by each PC


```{r calculate_propvar}
propvar_rlog_df <- tibble(
  PC = seq.int(1L, ncol(pca_rlog$x) ),
  prop_var = pca_rlog$sdev^2/sum(pca_rlog$sdev^2)
)
```

## Plot PCA percentage variance explained to find how many PCs to pay attention to.

```{r plot_percentvar, fig.width = 3, fig.height = 2}
plot_percentvar_rlog <- 
  ggplot(data = propvar_rlog_df, 
         aes(x = PC, y = prop_var)) +
  geom_col(fill = "blue") +
  scale_x_continuous("principal component",
                     limits = c(0.4,10.6), 
                     # hack limits to include the bars and a little space.
                     breaks = 1L:10L,
                     expand = c(0,0)) + 
  scale_y_continuous("prop. of variance", expand = c(0,0))
plot_percentvar_rlog
```

## Plot PC1 vs PC2 with default values

EDIT: the colour and other aesthetics may need to be edited to highlight features of the data

```{r plot_PC12_default, fig.height=5, fig.width=6, eval = FALSE}
ggplot(data = pcdf_rlog,
       aes(colour = interaction(Temperature,Media), 
           shape = FLC1)
       ) +
  scale_shape_manual(values = c(2, 20)) +
  scale_colour_brewer("Temp., Media", 
                      type = "qual", palette = 2) + 
  geom_point(aes(x = PC1, y = PC2))
```

## Plot PC1 vs PC2, 3, 4, also with default values

```{r plot_PC1234_text, fig.height=8, fig.width=8}
# This could be wraped in a function
pc_baseplot <- 
  ggplot(data = pcdf_rlog,
       aes(colour = interaction(Temperature,Media), 
           shape = FLC1)
       ) +
  scale_shape_manual(values = c(2, 20)) +
  scale_colour_brewer("Temp., Media", 
                      type = "qual", palette = 2) + 
  geom_hline(yintercept = 0, size = 0.5, colour = "grey80") +
  geom_vline(xintercept = 0, size = 0.5, colour = "grey80") +
  theme(legend.position = "none")

pcplot_12 <- pc_baseplot + geom_point(aes(x = PC1, y = PC2))
pcplot_13 <- pc_baseplot + geom_point(aes(x = PC1, y = PC3))
pcplot_14 <- pc_baseplot + geom_point(aes(x = PC1, y = PC4))
pcplot_23 <- pc_baseplot + geom_point(aes(x = PC2, y = PC3))
pcplot_24 <- pc_baseplot + geom_point(aes(x = PC2, y = PC4))
pcplot_34 <- pc_baseplot + geom_point(aes(x = PC3, y = PC4))

pcplot_legend <- get_legend(pcplot_12 + theme(legend.position = "right", legend.margin = margin(t = 0, r = 0, b = 20, l = 40, unit = "pt")))

plot_grid(pcplot_12, pcplot_legend, plot_percentvar_rlog,
          pcplot_13, pcplot_23, NULL, 
          pcplot_14, pcplot_24, pcplot_34,
          axis = "tblr", align = "hv")
```

# Session Info

Package versions, etc., that ran this file.

```{r session_info,results="show",message=TRUE}
sessionInfo()
```
